### ISSUE TITLE
Option To Add Multiple Categories To One Product

### ISSUE THREAD

**Issue by artistro08** [OFFLINE-GmbH/oc-mall-plugin#135]
Basically, I want the option to add more categories to one product. Say for instance I have a product called "Red Tee" and I want to add that product to three categories:

- Men
- Women
- Red

Is this possible now?

**Comment by tobias-kuendig** [comment]
No, currently not. This is a decision we made to not further complicate the filter system since it would become very hard to keep everything together when a product can belong to multiple categories that have multiple properties defined.

We will, however, introduce another feature that will make these setups possible. I just cannot tell you right now when it will be available.

**Comment by artistro08** [comment]
Thank you for the response. Much appreciated. I'll wait patiently :)

**Comment by tobias-kuendig** [comment]
It turns out, a lot has changed since this decision was made :grin: 	

I might be able to release a version that supports a `belongsToMany` relationship for Categories today.

**Comment by amdad** [comment]
wow!

**Comment by artistro08** [comment]
You'd be awesome if you did! 😁

**Comment by tobias-kuendig** [comment]
This feature is now available on the `develop` branch or via the edge updates as `1.1.0`. I'd like to test it a bit further until it gets released.

Please report any issues you encounter.

**Comment by artistro08** [comment]
My first question is, How do I access the category? I typed in on my CMS:

`/category/season-1` but it doesn't come up. Do I have to do a category hierarchy?

**Comment by artistro08** [comment]
So, The categories work! 🎉  But there is one thing I noticed. I had to update the products somehow before they got recognized that they were in a category. Other then that, it's working beautifully.

**Comment by tobias-kuendig** [comment]
Were they not visible in the frontend? But the category was selected in the backend?

**Comment by artistro08** [comment]
Yes.

**Comment by tobias-kuendig** [comment]
Is there anything in your `system.log`? They should have been re-indexed automatically during the update.

**Comment by artistro08** [comment]
Actually, yes: 
```
Illuminate\Database\Eloquent\RelationNotFoundException: Call to undefined relationship [category] on model [OFFLINE\Mall\Models\Product]. in /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Database/Eloquent/RelationNotFoundException.php:34Stack trace:
#0 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(548): Illuminate\Database\Eloquent\RelationNotFoundException::make(Object(OFFLINE\Mall\Models\Product), 'category')#1 [internal function]: Illuminate\Database\Eloquent\Builder-&gt;Illuminate\Database\Eloquent\{closure}()
#2 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Relations/Relation.php(89): call_user_func(Object(Closure))
#3 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(550): Illuminate\Database\Eloquent\Relations\Relation::noConstraints(Object(Closure))
#4 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(518): Illuminate\Database\Eloquent\Builder-&gt;getRelation('category')
#5 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(498): Illuminate\Database\Eloquent\Builder-&gt;eagerLoadRelation(Array, 'category', Object(Closure))
#6 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(466): Illuminate\Database\Eloquent\Builder-&gt;eagerLoadRelations(Array)
#7 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Database/Concerns/BuildsQueries.php(77): Illuminate\Database\Eloquent\Builder-&gt;get(Array)
#8 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(410): Illuminate\Database\Eloquent\Builder-&gt;first(Array)#9 /home/styleclo/public_html/styleclothingco/plugins/offline/mall/components/Product.php(370): Illuminate\Database\Eloquent\Builder-&gt;firstOrFail()
#10 /home/styleclo/public_html/styleclothingco/plugins/offline/mall/components/Product.php(321): OFFLINE\Mall\Components\Product-&gt;getProduct()
#11 /home/styleclo/public_html/styleclothingco/plugins/offline/mall/components/Product.php(187): OFFLINE\Mall\Components\Product-&gt;getItem()
#12 /home/styleclo/public_html/styleclothingco/modules/cms/classes/Controller.php(1464): OFFLINE\Mall\Components\Product-&gt;init()
#13 /home/styleclo/public_html/styleclothingco/modules/cms/classes/Controller.php(660): Cms\Classes\Controller-&gt;addComponent('product', 'product', Array)
#14 /home/styleclo/public_html/styleclothingco/modules/cms/classes/Controller.php(333): Cms\Classes\Controller-&gt;initComponents()
#15 /home/styleclo/public_html/styleclothingco/modules/cms/classes/Controller.php(224): Cms\Classes\Controller-&gt;runPage(Object(Cms\Classes\Page))#16 /home/styleclo/public_html/styleclothingco/modules/cms/classes/CmsController.php(50): Cms\Classes\Controller-&gt;run('product/bear-he...')
#17 [internal function]: Cms\Classes\CmsController-&gt;run('product/bear-he...')
#18 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Routing/Controller.php(54): call_user_func_array(Array, Array)
#19 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Routing/ControllerDispatcher.php(45): Illuminate\Routing\Controller-&gt;callAction('run', Array)
#20 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Routing/Route.php(212): Illuminate\Routing\ControllerDispatcher-&gt;dispatch(Object(Illuminate\Routing\Route), Object(Cms\Classes\CmsController), 'run')
#21 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Routing/Route.php(169): Illuminate\Routing\Route-&gt;runController()
#22 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Routing/Router.php(658): Illuminate\Routing\Route-&gt;run()
#23 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(30): Illuminate\Routing\Router-&gt;Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#24 /home/styleclo/public_html/styleclothingco/plugins/rainlab/translate/classes/LocaleMiddleware.php(29): Illuminate\Routing\Pipeline-&gt;Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#25 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(149): RainLab\Translate\Classes\LocaleMiddleware-&gt;handle(Object(Illuminate\Http\Request), Object(Closure))#26 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#27 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Routing/Middleware/SubstituteBindings.php(41): Illuminate\Routing\Pipeline-&gt;Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#28 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(149): Illuminate\Routing\Middleware\SubstituteBindings-&gt;handle(Object(Illuminate\Http\Request), Object(Closure))
#29 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#30 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/View/Middleware/ShareErrorsFromSession.php(49): Illuminate\Routing\Pipeline-&gt;Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#31 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(149): Illuminate\View\Middleware\ShareErrorsFromSession-&gt;handle(Object(Illuminate\Http\Request), Object(Closure))
#32 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#33 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Session/Middleware/StartSession.php(63): Illuminate\Routing\Pipeline-&gt;Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#34 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(149): Illuminate\Session\Middleware\StartSession-&gt;handle(Object(Illuminate\Http\Request), Object(Closure))
#35 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#36 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/AddQueuedCookiesToResponse.php(37): Illuminate\Routing\Pipeline-&gt;Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#37 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(149): Illuminate\Cookie\Middleware\AddQueuedCookiesToResponse-&gt;handle(Object(Illuminate\Http\Request), Object(Closure))
#38 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#39 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/EncryptCookies.php(66): Illuminate\Routing\Pipeline-&gt;Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#40 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(149): Illuminate\Cookie\Middleware\EncryptCookies-&gt;handle(Object(Illuminate\Http\Request), Object(Closure))
#41 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#42 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(102): Illuminate\Routing\Pipeline-&gt;Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#43 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Routing/Router.php(660): Illuminate\Pipeline\Pipeline-&gt;then(Object(Closure))
#44 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Routing/Router.php(635): Illuminate\Routing\Router-&gt;runRouteWithinStack(Object(Illuminate\Routing\Route), Object(Illuminate\Http\Request))
#45 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Routing/Router.php(601): Illuminate\Routing\Router-&gt;runRoute(Object(Illuminate\Http\Request), Object(Illuminate\Routing\Route))
#46 /home/styleclo/public_html/styleclothingco/vendor/october/rain/src/Router/CoreRouter.php(20): Illuminate\Routing\Router-&gt;dispatchToRoute(Object(Illuminate\Http\Request))
#47 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(176): October\Rain\Router\CoreRouter-&gt;dispatch(Object(Illuminate\Http\Request))
#48 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(30): Illuminate\Foundation\Http\Kernel-&gt;Illuminate\Foundation\Http\{closure}(Object(Illuminate\Http\Request))
#49 /home/styleclo/public_html/styleclothingco/plugins/offline/speedy/classes/middleware/CDNMiddleware.php(40): Illuminate\Routing\Pipeline-&gt;Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#50 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(149): OFFLINE\Speedy\Classes\Middleware\CDNMiddleware-&gt;handle(Object(Illuminate\Http\Request), Object(Closure))
#51 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#52 /home/styleclo/public_html/styleclothingco/plugins/offline/speedy/classes/middleware/Http2Middleware.php(38): Illuminate\Routing\Pipeline-&gt;Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#53 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(149): OFFLINE\Speedy\Classes\Middleware\Http2Middleware-&gt;handle(Object(Illuminate\Http\Request), Object(Closure))
#54 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#55 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/CheckForMaintenanceMode.php(46): Illuminate\Routing\Pipeline-&gt;Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#56 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(149): Illuminate\Foundation\Http\Middleware\CheckForMaintenanceMode-&gt;handle(Object(Illuminate\Http\Request), Object(Closure))
#57 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#58 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(102): Illuminate\Routing\Pipeline-&gt;Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#59 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(151): Illuminate\Pipeline\Pipeline-&gt;then(Object(Closure))
#60 /home/styleclo/public_html/styleclothingco/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(116): Illuminate\Foundation\Http\Kernel-&gt;sendRequestThroughRouter(Object(Illuminate\Http\Request))
#61 /home/styleclo/public_html/styleclothingco/index.php(43): Illuminate\Foundation\Http\Kernel-&gt;handle(Object(Illuminate\Http\Request))
#62 {main}
```

**Comment by tobias-kuendig** [comment]
Ah, so it showed in the category overview (Products component) but not when viewing a product (Product component)?

This bug was fixed in 2aad5fb

**Comment by artistro08** [comment]
They showed in both after I updated something on the product. It's weird.

**Comment by tobias-kuendig** [comment]
Alright, thank you for testing this out. I removed the auto re-indexing of the products from the migration to prevent any surprises. 

With 1.1.1 there comes a new database-backed index that will need a manual run of `php artisan mall:reindex` anyways.

**Comment by artistro08** [comment]
That works! 

No problem. Anything to help. 

Also, Random comment, your plugin is 10x better than JKShop.

**Comment by artistro08** [comment]
Are you gonna run this command on product updates to automatically populate the database?

**Comment by tobias-kuendig** [comment]
If you change a product in the backend yes, the index is updated automatically. This time the manual run is only needed because of the complete change in structure for all products (multiple categories) and the change from the file-based index to the database.

**Comment by artistro08** [comment]
For me, it doesn't update. I have to run the command again for it to work

**Comment by tobias-kuendig** [comment]
This feature is available in v1.1
